tls/extensions/s2n_client_psk.c

```diff
@@ -0,0 +1,250 @@
+            conn->psk_params.chosen_psk_wire_index = wire_index;
+            return S2N_RESULT_OK;
+        }
+
+        wire_index++;
+    }
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_client_psk_recv_binder_list(struct s2n_connection *conn, struct s2n_blob *partial_client_hello,
+        struct s2n_stuffer *wire_binders_in)
+{
+    ENSURE_REF(conn);
+
+    uint8_t wire_index = 0;
+    while(s2n_stuffer_data_available(wire_binders_in)) {
+        uint8_t wire_binder_size;
+        GUARD_AS_RESULT(s2n_stuffer_read_uint8(wire_binders_in, &wire_binder_size));
+
+        uint8_t *wire_binder_data;
+        ENSURE_REF(wire_binder_data = s2n_stuffer_raw_read(wire_binders_in, wire_binder_size));
+
+        struct s2n_blob wire_binder;
+        GUARD_AS_RESULT(s2n_blob_init(&wire_binder, wire_binder_data, wire_binder_size));
+
+        if (wire_index == conn->psk_params.chosen_psk_wire_index) {
+            GUARD_AS_RESULT(s2n_psk_verify_binder(conn, conn->psk_params.chosen_psk,
+                    partial_client_hello, &wire_binder));
+            return S2N_RESULT_OK;
+        }
+        wire_index++;
+    }
+    BAIL(S2N_ERR_BAD_MESSAGE);
+}
+
+static S2N_RESULT s2n_client_psk_recv_identities(struct s2n_connection *conn, struct s2n_stuffer *extension)
+{
+    ENSURE_REF(conn);
+
+    uint16_t identity_list_size;
+    GUARD_AS_RESULT(s2n_stuffer_read_uint16(extension, &identity_list_size));
+
+    uint8_t *identity_list_data;
+    ENSURE_REF(identity_list_data = s2n_stuffer_raw_read(extension, identity_list_size));
+
+    struct s2n_blob identity_list_blob;
+    GUARD_AS_RESULT(s2n_blob_init(&identity_list_blob, identity_list_data, identity_list_size));
+
+    struct s2n_stuffer identity_list;
+    GUARD_AS_RESULT(s2n_stuffer_init(&identity_list, &identity_list_blob));
+    GUARD_AS_RESULT(s2n_stuffer_skip_write(&identity_list, identity_list_blob.size));
+
+    return s2n_client_psk_recv_identity_list(conn, &identity_list);
+}
+
+static S2N_RESULT s2n_client_psk_recv_binders(struct s2n_connection *conn, struct s2n_stuffer *extension)
+{
+    uint16_t binder_list_size;
+    GUARD_AS_RESULT(s2n_stuffer_read_uint16(extension, &binder_list_size));
+
+    uint8_t *binder_list_data;
+    ENSURE_REF(binder_list_data = s2n_stuffer_raw_read(extension, binder_list_size));
+
+    struct s2n_blob binder_list_blob;
+    GUARD_AS_RESULT(s2n_blob_init(&binder_list_blob, binder_list_data, binder_list_size));
+
+    struct s2n_stuffer binder_list;
+    GUARD_AS_RESULT(s2n_stuffer_init(&binder_list, &binder_list_blob));
+    GUARD_AS_RESULT(s2n_stuffer_skip_write(&binder_list, binder_list_blob.size));
+
+    /* Record the ClientHello message up to but not including the binder list.
+     * This is required to calculate the binder for the chosen PSK. */
+    struct s2n_blob partial_client_hello = { 0 };
+    const struct s2n_stuffer *client_hello = &conn->handshake.io;
+    uint16_t partial_client_hello_size = client_hello->write_cursor
```

Comment:
if `write_cursor` is 0, you might have an arithmetic underflow here. You might add a check before this operation to make sure it doesn't lead to undefined behaviors.

Reply:
For example (you can probably write it better than that)
```c
ENSURE(SIZE_OF_BINDER_LIST_SIZE >= binder_list_blob.size, S2N_ERR_INTEGER_OVERFLOW);
ENSURE((binder_list_blob.size - SIZE_OF_BINDER_LIST_SIZE) < client_hello->write_cursor, S2N_ERR_INTEGER_OVERFLOW);
    uint16_t partial_client_hello_size = (client_hello->write_cursor
            - (binder_list_blob.size - SIZE_OF_BINDER_LIST_SIZE));
```